### PR TITLE
Allow for system format modules to be named. Fixes #2027

### LINKED ISSUE #2027: Cannot name system.register module
First off, thanks @guybedford for implementing System.register format, and thanks rollup maintainers for merging it in. We are experimenting with switching off of `jspm bundle` to `rollup -f system` and are excited about some of the possibilities with it!

The System.register() function has [an optional first argument string for naming the module](https://github.com/ModuleLoader/es-module-loader/blob/master/docs/system-register.md#bundled-vs-on-demand), but that is not currently supported by rollup. The string module name seems to correspond nicely with rollups `output.name` concept, since this is similar to a UMD or iife bundle name.

I propose that when rollup's `output.name` option is present, that it should be the name of the System.register module. I've got a pr coming soon that will do exactly that.

### PR DESCRIPTION
**Pull request by joeldenning**
See #2027. This makes it possible to create named System.register bundles

### CODE REVIEW

File: test/form/index.js [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
+const fs = require('fs');
 const assert = require('assert');
 const sander = require('sander');
 const rollup = require('../../dist/rollup');
 const { extend, loadConfig, normaliseOutput } = require('../utils.js');
 
 const samples = path.resolve(__dirname, 'samples');
 
-const FORMATS = ['amd', 'cjs', 'es', 'iife', 'umd'];
+const FORMATS = ['amd', 'cjs', 'es', 'iife', 'umd', 'system'];
```
Comment (joeldenning) [inline test/form/index.js]: Let me know your thoughts on this -- the vast majority of the tests do not have an `_expected/system.js` file defined for the "form" tests. But I'm thinking that all future tests should.

File: test/form/index.js [hunk 2]
```diff
@@ -37,6 +38,11 @@ describe('form', () => {
 			const createBundle = () => promise || (promise = rollup.rollup(options));
 
 			FORMATS.forEach(format => {
+				const skipBecauseNoSystem = format === 'system' && !fs.existsSync(path.resolve(__dirname, samples, dir, '_expected', 'system.js'))
+				if (skipBecauseNoSystem) {
+					return;
```
Comment (joeldenning) [inline test/form/index.js]: Don't call `it()` for "form" tests that don't have a system format expectation.

This prevents breaking existing tests but also providing a consistent way to add systemjs expectations in future tests.

File: test/form/samples/output-named-library/_config.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+  description: 'names bundles correctly',
+  options: {
+    output: {
+      name: 'libraryName',
```
Comment (joeldenning) [inline test/form/samples/output-named-library/_config.js]: Before this pr, this name was ignored in the outputted System.register format.

File: test/form/samples/output-named-library/_expected/system.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+System.register('libraryName', [], function (exports, module) {
```
Comment (joeldenning) [inline test/form/samples/output-named-library/_expected/system.js]: Before my pr, `'libraryName', ` would not have appeared in this System.register call

File: test/form/index.js (hunk 1 above)
Comment (guybedford) [reply test/form/index.js]: This is included in the PR at https://github.com/rollup/rollup/pull/2022. Perhaps rebase to that if you want to do this?

File: test/form/index.js (hunk 2 above)
Comment (guybedford) [reply test/form/index.js]: This can also be avoided by rebasing to the other branch.

**Comment by guybedford** [comment]
Looking through the issue queue I can find nothing on named AMD defines, so assume it just hasn't been a needed use case. I take it code splitting is the feature that makes this different here really.

**Comment by joeldenning** [comment]
Thanks @guybedford for the review -- I've rebased to master and have updated all the tests that previously were not naming their System.register bundles.

**Comment by guybedford** [comment]
It's worth considering here that this won't apply to the code splitting workflow very well.

It might be nice to think up a solution that can work with both single-file and code splitting workflows instead of having separate mechanisms for both of these.

Perhaps combining with the chunk naming feature would make more sense for code splitting integration - https://github.com/rollup/rollup/issues/1995. Just thinking aloud here though...

**Comment by joeldenning** [comment]
@guybedford I'm not very familiar with rollup code splitting -- I heard it was announced pretty recently but don't know the mechanics of how it would work.

How do umd and iife handle named libraries and code splitting?

**Comment by lukastaegert** [comment]
@guybedford This PR is marked as requesting changes from you. Is this still the case or can this be merged?

**Comment by guybedford** [comment]
I still think it's worth discussing the API use cases here with regards to code splitting workflows a little further to ensure we have an approach that can work in both scenarios.

The new code splitting work will allow customizing all chunk names (in progress, I'm hoping to provide a PR shortly). If this is the case, it may well provide an API that can work for this feature being just a boolean flag.

In addition it may be worth making a separate top-level method for concatenation, something like `generateCombined` which only works for the `system` and `amd` formats (and in future if ES modules support named wrappers).

Then the single-file case becomes:

```js
rollup.rollup({
  // allow input object to work in single-file case for naming:
  input: { 
    'inputName': 'input/file/path.js'
  }
}).generate({
  chunkNames: 'chunk-[hash].js',
  format: 'system',
  combineNamedModules: true
});
```

Which would output a single source containing `System.register('inputName', [',/chunk-asdf897.js'], ...); System.register(',/chunk-asdf897.js', [], ...)`.

**Comment by lukastaegert** [comment]
Ok, so I'll leave this out of the current release.

**Comment by joeldenning** [comment]
@guybedford since it sounds like code splitting is potentially going to be overhauled in the near future, would it make sense to address this problem at that time? The `combineNamedModules` option seems really cool (and relevant to #2036!), but it seems like there is some discussion and work to be done to decide if it will even happen.

Getting named system register modules at all would be cool -- I vote for merging this in and then iterating on it once the code splits / combineNamedModules discussions are fleshed out.

**Comment by guybedford** [comment]
@joeldenning I'm aiming to get a PR for chunk naming very soon... once that is in we can build this feature on top of that I think.

**Comment by guybedford** [comment]
I just noticed there is actually an `amd.id` input option. It's probably worth ensuring that whatever is done here either deprecates that or works with it then.

File: src/finalisers/system.ts [hunk 5]
```diff
@@ -109,8 +111,9 @@ export default function system (
 			`\n${t}${varOrConst} _starExcludes = { ${Array.from(starExcludes).join(': 1, ')}${starExcludes.size ? ': 1' : ''} };`;
 
 	const importBindingsSection = importBindings.length ? `\n${t}var ${importBindings.join(', ')};` : '';
+	const registeredName = outputOptions.name ? `'${outputOptions.name}', ` : '';
```
Comment (lukastaegert) [inline src/finalisers/system.ts]: Looking through rollup's [documentation](https://rollupjs.org/guide/en#big-list-of-options), I think using `name` here is an abuse of this option which is meant to define global variable names for IIFE/umd bundles (maybe we want to add SystemJS to our umd output at some point?). As @guybedford pointed out, the `amd.id` option would be more equivalent except for the `amd` part of the name. So I would prefer deprecating `amd.id` in favour of a new output option. Not sure what the best name could be;  we could call it something like `moduleId` as it is the id of the generated module as used by the loader. Maybe @guybedford has a more elaborate opinion on this.

**Comment by guybedford** [comment]
Yup, we'd need to avoid the name overload. `moduleId` as a new option to replace `amd.id` sounds really sensible to me actually, with it only supported in the non-code-splitting build. For the code splitting case, `moduleId: true` could possibly be an option as well I guess.

**Comment by joeldenning** [comment]
@lukastaegert @guybedford I started work on switching over to `moduleId` instead of `name`, but discovered that [`moduleId` is already a rollup output option that means something else](https://github.com/rollup/rollup/blob/3d122d78f0c954340838efa0b17efd54dbc7acd1/src/rollup/index.ts#L170). On top of that, it's a deprecated option, so I'm not sure if we would want to piggy back on it.

Maybe `moduleName`?

Also, would you like me to deprecate `amd.id` in favor of the new property? I think I know how to do so in the amd finalizer, but might need some help hunting down all the places where deprecation warnings are supposed to happen / documentation needs to be updated.

**Comment by guybedford** [comment]
It seems `moduleId` was exactly for this purpose but deprecated in https://github.com/rollup/rollup/commit/979b5e4969244254d85c27b5dc068a0c8e8ffc48.

At the time, AMD was the only module format with a moduleId, so I see the reasoning, but perhaps now with both AMD and System formats, it would be worth reconsidering this.

If we do undeprecate moduleId, then we should ensure that we provide warnings when moduleId is used with a format that is not AMD or System.

Personally I'd be for deprecating `amd.id` for going back to `moduleId`, but open to suggestions. @Rich-Harris @lukastaegert what do you think?

**Comment by kerwitz** [comment]
What's the status here?

**Comment by CoskunSunali** [comment]
Looking forward for this PR to be merged when ready. Can you guys be so kind to update us with a status?

**Comment by lukastaegert** [comment]
I think the only reason this has remained unmerged was the choice of option name. But I guess using `name` here should be ok so I'll merge this now.

**Comment by lukastaegert** [comment]
Sorry for the long wait!

**Comment by lukastaegert** [comment]
Will be a few days until we do another release, though.

**Comment by CoskunSunali** [comment]
🕺 Better late than never!

Thanks for the update.

**Pull request merged**